src/rgw/rgw_rados.cc

```diff
@@ -6065,27 +6065,58 @@ int RGWRados::Bucket::UpdateIndex::guard_reshard(const DoutPrefixProvider *dpp,
       ldpp_dout(dpp, 5) << "failed to get BucketShard object: ret=" << ret << dendl;
       return ret;
     }
+
     r = call(bs);
-    if (r != -ERR_BUSY_RESHARDING) {
+    if (r != -ERR_BUSY_RESHARDING && r != -ENOENT) {
       break;
     }
-    ldpp_dout(dpp, 0) << "NOTICE: resharding operation on bucket index detected, blocking" << dendl;
-    string new_bucket_id;
-    r = store->block_while_resharding(bs, &new_bucket_id,
-                                      target->bucket_info, null_yield, dpp);
-    if (r == -ERR_BUSY_RESHARDING) {
-      continue;
-    }
-    if (r < 0) {
-      return r;
+
+    std::string new_bucket_id;
+
+    // different logic depending whether resharding completed or is
+    // underway
+
+    if (r == -ENOENT) { // case where resharding must have completed
+      ldpp_dout(dpp, 0) <<
+	"NOTICE: resharding operation recently completed, invalidating "
+	"old BucketInfo" << dendl;
+
+      RGWBucketInfo new_bucket_info;
+      r = store->fetch_new_bucket_id(target->bucket_info,
+				     "re-reading BucketInfo due to recent resharding",
+				     &new_bucket_info,
+				     new_bucket_id,
+				     dpp);
```

Comment:
you seem to have added the `RGWBucketInfo* save_bucket_info` argument for this caller, but it doesn't look like you're actually using this `new_bucket_info` anywhere?